// @langchain/google-webauth/dist/auth.js
import { getEnvironmentVariable } from "@langchain/core/utils/env";
import { ensureAuthOptionScopes, GoogleAbstractedFetchClient, } from "@langchain/google-common";
import { getAccessToken, getCredentials, } from "web-auth-library/google";
export class WebGoogleAuth extends GoogleAbstractedFetchClient {
    constructor(fields) {
        super();
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        const options = fields?.authOptions;
        const accessToken = options?.accessToken;
        const credentials = options?.credentials ??
            getEnvironmentVariable("GOOGLE_WEB_CREDENTIALS") ??
            getEnvironmentVariable("GOOGLE_VERTEX_AI_WEB_CREDENTIALS");
        if (credentials === undefined)
            throw new Error(`Credentials not found. Please set the GOOGLE_WEB_CREDENTIALS environment variable or pass credentials into "authOptions.credentials".`);
        this.options = ensureAuthOptionScopes({ ...options, accessToken, credentials }, "scope", fields?.platformType);
    }
    get clientType() {
        return "webauth";
    }
    async getProjectId() {
        const credentials = getCredentials(this.options.credentials);
        return credentials.project_id;
    }
    async request(opts) {
        let { accessToken } = this.options;
        if (accessToken === undefined) {
            accessToken = await getAccessToken(this.options);
        }
        const authHeader = {
            Authorization: `Bearer ${accessToken}`,
        };
        return this._request(opts?.url, opts, authHeader);
    }
}
